### PR TITLE
[feat] Add option to print performance report

### LINKED ISSUE #634: Add a 'benchmarking' mode in ReFrame
When performance tests are run in this mode (perhaps allow only performance tests or ignore non-performance tests), ReFrame will report the performance achieved in different performance tests.

### PR DESCRIPTION
**Pull request by rsarm**
Closes #634

### CODE REVIEW

File: reframe/core/pipeline.py [hunk 1]
```diff
@@ -1119,11 +1120,17 @@ def check_performance(self):
                         "tag `%s' not resolved in references for `%s'" %
                         (tag, self._current_partition.fullname))
 
-                perf_values.append((value, self.reference[key]))
+                self.perf_value = value
```
Comment (teojgo) [inline reframe/core/pipeline.py]: Why is this field needed?

File: reframe/core/pipeline.py [hunk 2]
```diff
@@ -1119,11 +1120,17 @@ def check_performance(self):
...
+                self.perf_name = key
+
+                self.perf_values.append((value, self.reference[key]))
+                self.perf_keys.append(key)
                 self._perf_logger.log_performance(logging.INFO, tag, value,
                                                   *self.reference[key])
 
-            for val, reference in perf_values:
+            # self.perf_values = perf_values
```
Comment (teojgo) [inline reframe/core/pipeline.py]: Remove these line

File: reframe/core/pipeline.py [hunk 3]
```diff
@@ -1119,11 +1120,17 @@ def check_performance(self):
                         "tag `%s' not resolved in references for `%s'" %
                         (tag, self._current_partition.fullname))
 
-                perf_values.append((value, self.reference[key]))
+                self.perf_value = value
+                self.perf_name = key
```
Comment (teojgo) [inline reframe/core/pipeline.py]: Same as above.

File: reframe/frontend/statistics.py [hunk 4]
```diff
@@ -107,3 +107,27 @@ def failure_report(self):
...
+                if type(p[1][-1]) is str:
+                    unit = p[1][-1]
+                else:
+                    unit = ''
+
+                report.append(' * [%s:%s] %s: %s %s' %
+                              (system, partition, key, p[0], unit))
+
+        report.append(line_width * '-')
+        return '\n'.join(report)
```
[the remark this answers is not in the document]
Comment (vkarak) [reply reframe/frontend/statistics.py]: This is my proposal:

```
==============================================================================
PERFORMANCE REPORT
------------------------------------------------------------------------------
StreamTest
- dom:gpu
   - PrgEnv-cray
      * triad: 56913.7 MB/s
   - PrgEnv-gnu
      * triad: 44166.1 MB/s
   - PrgEnv-intel
      * triad: 58853.7 MB/s
- dom:mc
   - PrgEnv-cray
      * triad: 55364.3 MB/s
   - PrgEnv-gnu
      * triad: 43795.2 MB/s
   - PrgEnv-intel
      * triad: 55007.3 MB/s
------------------------------------------------------------------------------
HaloCellExchangeTest
- dom:gpu
   - PrgEnv-cray
      * time_2_10: 2.771997e-06 s
      * time_2_10000: 9.388208e-06 s
      * time_2_1000000: 0.0004652247 s
      * time_4_10: 4.093409e-06 s
      * time_4_10000: 1.484702e-05 s
      * time_4_1000000: 0.0007117964 s
      * time_6_10: 5.499196e-06 s
      * time_6_10000: 1.900601e-05 s
      * time_6_1000000: 0.0008331874 s
   - PrgEnv-gnu
      * time_2_10: 2.848721e-06 s
      * time_2_10000: 9.4347e-06 s
      * time_2_1000000: 0.0004597242 s
      * time_4_10: 4.120207e-06 s
      * time_4_10000: 1.53404e-05 s
      * time_4_1000000: 0.0007522151 s
      * time_6_10: 5.907083e-06 s
      * time_6_10000: 1.966281e-05 s
      * time_6_1000000: 0.0008944393 s
------------------------------------------------------------------------------
```

File: reframe/frontend/statistics.py [hunk 5]
```diff
@@ -107,3 +107,34 @@ def failure_report(self):
...
+                    report.append('%s' % t.check.name)
+                    previous_name = t.check.name
+
+                if t.check.current_partition.fullname != previous_part:
+                    report.append('- %s' % t.check.current_partition.fullname)
+                    previous_part = t.check.current_partition.fullname
+
+                report.append('   - %s' % t.check.current_environ)
+
+            for k, p in t.check.perfvalues.items():
```
Comment (vkarak) [inline reframe/frontend/statistics.py]:
```suggestion
            for key, ref in t.check.perfvalues.items():
```

File: reframe/frontend/statistics.py [hunk 6]
```diff
@@ -107,3 +107,34 @@ def failure_report(self):
...
+                key = k.split(':')[-1]
```
Comment (vkarak) [inline reframe/frontend/statistics.py]:
```suggestion
                var = k.split(':')[-1]
```

File: reframe/frontend/statistics.py [hunk 7]
```diff
@@ -107,3 +107,34 @@ def failure_report(self):
...
+                if type(p[1][-1]) is str:
+                    unit = p[1][-1]
+                else:
+                    unit = ''
```
Comment (vkarak) [inline reframe/frontend/statistics.py]: val = p[0]
try:
    unit = ref[3]
except IndexError:
    unit = '(no unit specified)'

File: reframe/frontend/statistics.py [hunk 8]
```diff
@@ -107,3 +107,34 @@ def failure_report(self):
...
+
+                report.append('      * %s: %s %s' % (key, p[0], unit))
```
Comment (vkarak) [inline reframe/frontend/statistics.py]:
```suggestion
                report.append('      * %s: %s %s' % (var, val, unit))
```

**Comment by rsarm** [comment]
@jenkins-cscs retry kesch

**Comment by vkarak** [comment]
@jenkins-cscs retry dom

**Pull request merged**